linux_rootfs.go

```diff
@@ -0,0 +1,407 @@
+// +build linux
+
+package libcontainer
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/docker/docker/pkg/symlink"
+	"github.com/docker/libcontainer/configs"
+	"github.com/docker/libcontainer/label"
+)
+
+const defaultMountFlags = syscall.MS_NOEXEC | syscall.MS_NOSUID | syscall.MS_NODEV
+
+type mount struct {
+	source string
+	path   string
+	device string
+	flags  int
+	data   string
+}
+
+// setupRootfs sets up the devices, mount points, and filesystems for use inside a
+// new mount namespace.
+func setupRootfs(config *configs.Config) (err error) {
+	if err := prepareRoot(config); err != nil {
+		return err
+	}
+	if err := mountSystem(config); err != nil {
+		return err
+	}
+	// apply any user specified mounts within the new mount namespace
+	for _, m := range config.Mounts {
+		if err := mountUserMount(m, config.Rootfs, config.MountLabel); err != nil {
+			return err
+		}
+	}
+	if err := createDevices(config); err != nil {
+		return err
+	}
+	if err := setupPtmx(config); err != nil {
+		return err
+	}
+	uid, err := config.HostUID()
+	if err != nil {
+		return err
+	}
+	// stdin, stdout and stderr could be pointing to /dev/null from parent namespace.
+	// Re-open them inside this namespace.
+	if uid == 0 {
```

Comment:
Here was a comment "// FIXME: Need to fix this for user namespaces" and we check uid to detect USERNS. Currently we can use a proper check here:
if !config.Namespaces.Contains(configs.NEWUSER) {